core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2383,23 +2383,41 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
 
       def nextSize(sourceSize: Int): Int = (factor * sourceSize).toInt
 
-      def go(acc: Chunk[O], sizeOpt: Int, s: Pull[F2, O, Unit]): Pull[F2, O, Unit] =
-        s.uncons.flatMap {
-          case Some((hd, tl)) =>
-            val size = if (sizeOpt > 0) sizeOpt else nextSize(hd.size)
-            if (acc.size < size)
-              go(acc ++ hd, size, tl)
-            else if (acc.size == size)
-              Pull.output(acc) >> go(hd, size, tl)
-            else {
-              val (out, rem) = acc.splitAt(size - 1)
-              Pull.output(out) >> go(rem ++ hd, -1, tl)
+      def go(
+          acc: Chunk[O],
+          sizeOpt: Int,
+          lastChunkSize: Int,
+          s: Pull[F2, O, Unit]
+      ): Pull[F2, O, Unit] = {
+
+        val size = if (sizeOpt > 0) sizeOpt else nextSize(lastChunkSize)
+
+        if (acc.size < size)
+          s.uncons.flatMap {
+            case None => Pull.output(acc)
+            case Some((hd, tl)) =>
+              go(acc ++ hd, size, hd.size, tl)
+          }
+        else if (acc.size == size)
+          Pull.output(acc) >>
+            s.uncons.flatMap {
+              case None => Pull.done
+              case Some((hd, tl)) =>
+                go(hd, size, hd.size, tl)
             }
-          case None =>
-            Pull.output(acc)
+        else {
+          val (out, rem) = acc.splitAt(size - 1)
+          Pull.output(out) >> go(rem, -1, lastChunkSize, s)
+
         }
 
-      go(Chunk.empty, -1, underlying).stream
+      }
+
+      underlying.uncons.flatMap {
+        case None => Pull.done
```

Comment:
Idea, since the new implementation is slightly more complex anyway, maybe we could avoid creation of the `random` object (`val random = new scala.util.Random(seed)`) in this empty case, since it's then not used? Or maybe it's not worth it, not sure. :slightly_smiling_face:

Reply:
We could, but I'm not sure if it's worth it. It'd only have any effect if the random rechunking was called repeatedly on empty streams, which seems like an extraordinarily niche scenario.

Anyway, I'm happy to do it if told otherwise, but I'll save the CI electricity for now